### PR TITLE
feat: add basic cli

### LINKED ISSUE #84: feat: add basic CLI
Add a basic CLI skeleton which allows to:
- Run a Cairo program given its compiled `.json`
- Print the memory
- Print the relocated memory
- Export the relocated memory
- Export the trace

As seen in #60, the relocated memory starts at address 1 for an implementation-specific reason. The `RunOptions` config defaults `relocateOffset` to 0, it should be the same for the CLI. However, for ease of use and expectations from the user (behaviour in other VMs), it might be better to default the `relocateOffset` to 1 in the CLI and add a flag to modify its value

### PR DESCRIPTION
**Pull request by zmalatrax**
Closes #84 

Implements a basic CLI to run the cairo vmin TypeSscript

Dependencies:
- [`commander.js`](https://github.com/tj/commander.js) to build the CLI
- [`consola`](https://github.com/unjs/consola) for an elegant console wrapper

Design choices:
Separate the proof mode and the execution mode by two subcommands rather than a flag
- `cairo-vm-ts run <program.json>` in execution mode (default command so directly using `cairo-vm-ts` will trigger the `run` subcommand)
- `cairo-vm-ts prove <program.json>` in proof mode (TBD)

### CODE REVIEW

**Comment by zmalatrax** [comment]
Detailed list of the implemented flags in this [comment](https://github.com/kkrt-labs/cairo-vm-ts/issues/83#issuecomment-2160306432)

File: README.md [hunk 1]
```diff
@@ -92,13 +93,39 @@ performance and safety. While the ones of our TypeScript implementation is
...
+
+### CLI
+
+You can install the CLI `cairo-vm-ts` by doing the following:
+
+1. Clone this repo: `git clone git@github.com:kkrt-labs/cairo-vm-ts.git`
+2. Go to the cloned directory: `cd cairo-vm-ts`
+3. Install the dependencies: `bun install`
+4. Register the package as a _linkable_ package: `bun link`
+5. Link the package: `bun link cairo-vm-ts`
```
Comment (ClementWalter) [inline README.md:106]: link two times

File: README.md [hunk 2]
```diff
@@ -92,13 +93,39 @@ performance and safety. While the ones of our TypeScript implementation is
...
+
+Example usage:
+
+```bash
+cairo-vm-ts run fibonacci.json --export-memory fib_mem.bin --print-memory --print-output
```
Comment (ClementWalter) [inline README.md:111]: should we keep only `cairo` ? like `cairo run ...`

File: bin/cli.ts [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+  .option(
+    '--export-memory <MEMORY_FILENAME>',
+    'export the relocated memory, little-endian encoded'
+  )
+  .option('--print-trace', 'print the trace')
+  .option('--print-memory', 'print the non-relocated memory')
+  .option('--print-relocated-memory', 'print the relocated memory')
+  .option('--print-output', 'print the output segment')
+  .action(async (path, options) => {
+    if (!path.match(/\.json$/)) throw new Error('Provided file is not a JSON');
```
Comment (ClementWalter) [inline bin/cli.ts:47]: I think that this should be done in a file parser and not in the body

File: bin/cli.ts [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+      process.exit(1);
+    }
+
+    consola.info(`Cairo VM TS ${VERSION_CLI} - Execution Mode`);
+
+    const program = parseProgram(fs.readFileSync(path, 'utf-8'));
+    const runner = new CairoRunner(program);
+    const config: RunOptions = {
+      relocate: relocate,
+      relocateOffset: offset,
```
Comment (ClementWalter) [inline bin/cli.ts:77]: can't we keep the same naming ?

File: bin/cli.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+consola.options = {
+  ...consola.options,
+  formatOptions: {
+    date: false,
+  },
+};
+
+const VERSION_CLI = '0.1.0';
+
+const program = new Command().name('cairo-vm-ts').version(VERSION_CLI);
```
[the remark this answers is not in the document]
Comment (zmalatrax) [reply bin/cli.ts]: The value in `name` will be the one displayed in the help section 'Usage', so it must match the key in `bin` object from `package.json`

It currently is `cairo-vm-ts` but can be renamed to `cairo` for example

File: bin/cli.ts (hunk 4 above)
Comment (zmalatrax) [reply bin/cli.ts]: Sure, but we'll need to either rename the flag `--offset` to `--relocate-offset` with is kinda long imo, or rename the `relocateOffset` property to `offset`. I think renaming `relocateOffset` is ok

File: bin/cli.ts (hunk 3 above)
Comment (zmalatrax) [reply bin/cli.ts]: moved to an `argParser` on the argument declaration

File: package.json [hunk 6]
```diff
@@ -2,18 +2,29 @@
...
   "author": "Clément Walter <clement@kakarot.org>",
   "license": "Apache-2.0",
-  "devDependencies": {
-    "bun-types": "^1.1.12",
-    "typescript": "^5.2.2"
+  "main": "dist/index.js",
+  "types": "dist/types/index.d.ts",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "start": "echo \"Error: no entrypoint yet\" && exit 1",
```
Comment (ClementWalter) [inline package.json:12]:
```suggestion
```

File: src/cli.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    consola.info(`Cairo VM TS ${VERSION_CLI} - Execution Mode`);
+
+    const program = parseProgram(fs.readFileSync(String(path), 'utf-8'));
+    const runner = new CairoRunner(program);
+    const config: RunOptions = {
+      relocate: relocate,
+      offset: offset,
+    };
```
Comment (ClementWalter) [inline src/cli.ts:87]:
```suggestion
    const config: RunOptions = { relocate, offset };
```

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**